### PR TITLE
refactored sampler class

### PR DESCRIPTION
**Pull request by gcskoenig**
First version of the sampler refactoring. I moved as much of the code as possible into the Sampler class.

Furthermore, while iterating the code I came to the conclusion that is misleading to implement train(J, G) with J being multivariate as sampling each column individually. It may be ok for our standard purposes, but I am more and more convinced that it would be useful to resample multiple variables at once. Then, their covariance structure has to be respected as well, of course.

Furthermore I removed automatic training if the sampler is not trained yet in the sampler function -- I feel the user will be more aware of whats happening when training has to happen outside the sampler class.

For Multivariate gaussian distribution I know that this functionality is implemented. I don't remember for CNF. Is a call with a multivariate J allowed? @Valentyn1997

Next steps

- [x] Test compatibility with Explainer functions
- [ ] Test compatibility with CNF sampler
- [x] raise exception when sampler not trained
- [x] discussion on logging vs printing

Not ready to merge obviously

### CODE REVIEW

File: tests/test_gaussian_sampling.py [hunk 1]
```diff
@@ -15,14 +15,19 @@
...
 
     def compute_sample(self, J, G):
         X_G = X_test[:, G]
         X_J = X_test[:, J]
 
+        gaussian_estimator = GaussianConditionalEstimator()
+        gaussian_estimator.fit(train_inputs=X_train[:, J], train_context=X_train[:, G])
+        X_J_recovered = gaussian_estimator.sample(X_test[:, G])
+
         sample = train_gaussian(X_train, J, G)
```
Comment (Valentyn1997) [inline tests/test_gaussian_sampling.py:31]: train_gaussian is deprecated?

**Review by Valentyn1997 — COMMENTED** [review]
I reviewed this code and added comments. 

You could check the compatibility with Explainer functions/ CNF sampler with **examples/chains/test_samplers.py.**

I guess, the logging is more preferable - if you want to change verbosity, you just locally change the logging level.

P.S. At the moment, CNFSampler is not ready to the multivariate J. I will need to write extra code for inverse Radial flow in this case.

**Comment by Valentyn1997** [comment]
@gcskoenig - you could also add all the tests to the /.github/workflows/python-package.yml - so that it will be automatically checked after each git push

File: rfi/samplers/sampler.py [hunk 2]
```diff
@@ -55,7 +106,8 @@ def train(self, J, G):
             Nothing. Now the sample function can be used
             to resample on seen or unseen data.
         """
-        pass
+        print('Training Sampler for: {} | {}'.format(J, G))
```
[the remark this answers is not in the document]
Comment (gcskoenig) [reply rfi/samplers/sampler.py:109]: https://docs.python.org/3/howto/logging.html

I don't see logger info as an replacement here. I would like this message to be printed anytime the training of a sampler is run.

File: rfi/samplers/sampler.py [hunk 3]
```diff
@@ -69,4 +121,16 @@ def sample(self, X_test, J, G):
...
         """
-        pass
+                # initialize numpy matrix
+        J = np.array(J, dtype=np.int16)
+        sampled_data = np.zeros((X_test.shape[0], J.shape[0]))
+
+        # sample
+        G_key, J_key = Sampler._to_key(G), Sampler._to_key(J)
+
+        if not super().is_trained(J, G):
```
[the remark this answers is not in the document]
Comment (gcskoenig) [reply rfi/samplers/sampler.py:131]: Copied the code from the subclass. Good catch

File: rfi/samplers/cnflow.py [hunk 4]
```diff
@@ -17,35 +17,16 @@ def __init__(self, X_train):
         super().__init__(X_train)
         self.time_budget_s = 120
 
-    def _train_j(self, j, G):
+    def train(self, J, G, verbose=verbose):
```
[the remark this answers is not in the document]
Comment (gcskoenig) [reply rfi/samplers/cnflow.py:20]: Resolved

File: tests/test_gaussian_sampling.py (hunk 1 above)
Comment (gcskoenig) [reply tests/test_gaussian_sampling.py:31]: and replaced with the lines above

**Comment by gcskoenig** [comment]
As mentioned earlier those tests currently fail, so I did not include them into CI.
I would appreciate if we could write all our tests in this format in the future.

File: rfi/samplers/sampler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Valentyn1997) [reply rfi/samplers/sampler.py:109]: Ah, ok

**Comment by Valentyn1997** [comment]
> As mentioned earlier those tests currently fail, so I did not include them into CI.
> I would appreciate if we could write all our tests in this format in the future.

Yeah, I will need to rewrite examples/chains/test_samplers.py

**Comment by gcskoenig** [comment]
The code is now ready to merge from my side.  Have not tested compatiblity with CNF, have not removed Distribution heritage. Could you review? Would suggest to introduce CNF compatibility later on.

**Comment by Valentyn1997** [comment]
I reviewed, looks really nice!

**Review by Valentyn1997 — APPROVED** [review]

**Pull request merged**